### PR TITLE
refactor: Simplify command state update methods

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

Address comments raised on https://github.com/thin-edge/thin-edge.io/pull/3307

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/pull/3307

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 1]
```diff
@@ -279,16 +275,12 @@ impl WorkflowRepository {
             if let Err(err) = tokio::fs::copy(source.clone(), target.clone()).await {
                 error!("Fail to persist a copy of {source} as {target}: {err}");
             } else {
-                self.in_use_copies.insert(version.clone(), 1);
+                self.in_use_copies.insert(version.to_string(), 1);
```
Comment (jarhodes314) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:278]: Minor, `to_owned` better communicates the intent here
```suggestion
                self.in_use_copies.insert(version.to_owned(), 1);
```

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 2]
```diff
@@ -136,10 +136,10 @@ impl WorkflowSupervisor {
...
     /// Return the current version if any.
     pub fn use_current_version(&mut self, operation: &OperationName) -> Option<WorkflowVersion> {
-        match self.workflows.get_mut(&operation.as_str().into()) {
-            Some(versions) => versions.use_current_version().cloned(),
-            None => None,
-        }
+        self.workflows
+            .get_mut(&operation.as_str().into())
+            .map(WorkflowVersions::use_current_version)?
+            .cloned()
```
Comment (jarhodes314) [inline crates/core/tedge_api/src/workflow/supervisor.rs:142]: This function could also take `&str`, and we could also implement `Borrow<str>` for `OperationType`, which would save us the conversion into `&OperationType` when we query `self.workflows`.

The `.map()?` is also an interesting pattern. I think either of these are clearer:
```suggestion
        self.workflows
            .get_mut(&operation.as_str().into())
            .and_then(WorkflowVersions::use_current_version)
            .cloned()
```

```suggestion
        self.workflows
            .get_mut(&operation.as_str().into())?
            .use_current_version()
            .cloned()
```

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 3]
```diff
@@ -410,14 +402,14 @@ impl WorkflowRepository {
         operation: &OperationType,
         command_state: GenericCommandState,
     ) -> Result<Option<GenericCommandState>, WorkflowExecutionError> {
+        let operation_name = operation.to_string();
```
Comment (jarhodes314) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:405]: Minor: I think `to_owned` better signifies the intent here
```suggestion
        let operation_name = operation.to_owned();
```

File: crates/core/tedge_api/src/workflow/state.rs [hunk 4]
```diff
@@ -186,8 +186,18 @@ impl GenericCommandState {
...
-    pub fn update_with_key_value(self, key: &str, val: &str) -> Self {
-        self.update_with_json(json!({ key: val }))
+    pub fn with_key_value(mut self, key: &str, val: &str) -> Self {
+        self.set_key_value(key, val);
+        self
+    }
+
+    fn set_key_value(&mut self, key: &str, val: &str) {
+        if let Some(o) = self.payload.as_object_mut() {
+            o.insert(key.to_string(), val.into());
```
Comment (jarhodes314) [inline crates/core/tedge_api/src/workflow/state.rs:196]: Again, minor:

```suggestion
            o.insert(key.into(), val.into());
```

File: crates/core/tedge_api/src/workflow/state.rs [hunk 5]
```diff
@@ -186,8 +186,18 @@ impl GenericCommandState {
...
+        }
+        if key == STATUS {
+            self.status = val.to_string();
```
Comment (jarhodes314) [inline crates/core/tedge_api/src/workflow/state.rs:199]:
```suggestion
            self.status = val.to_owned();
```

**Review by jarhodes314 — APPROVED** [review]

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: Done: https://github.com/thin-edge/thin-edge.io/pull/3309/commits/eecd22274eb3baed62ffefa53be54d1768dc1c99

File: crates/core/tedge_api/src/workflow/state.rs (hunk 4 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/state.rs]: Done: https://github.com/thin-edge/thin-edge.io/pull/3309/commits/eecd22274eb3baed62ffefa53be54d1768dc1c99

File: crates/core/tedge_api/src/workflow/state.rs (hunk 5 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/state.rs]: Done: https://github.com/thin-edge/thin-edge.io/pull/3309/commits/eecd22274eb3baed62ffefa53be54d1768dc1c99

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs]: I opted for the 2nd proposal: https://github.com/thin-edge/thin-edge.io/pull/3309/commits/eecd22274eb3baed62ffefa53be54d1768dc1c99

Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs]: > This function could also take &str, and we could also implement Borrow<str> for OperationType, which would save us the conversion into &OperationType when we query self.workflows.

`Borrow<str>` cannot be implemented neither by `OperationType` (which is an enum with no string carrier) nor by `OperationName` (which is an alias).

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: `to_owned()` is not appropriate here.

To convey the intent, I replaced `to_string()` by `name()`: https://github.com/thin-edge/thin-edge.io/pull/3309/commits/33268b91f4e499a9bb1cc2b9c79d7886680f270e

**Pull request merged**